Added Ramulator2 Integration

maccoymerrell opened the pull request.

maccoymerrell committed:
Fixed bug with cache line dirty flag being reset.

maccoymerrell committed:
Added refresh and minor timing corrections to DRAM

maccoymerrell committed:
Merge branch 'develop' into DRAM_Controller

maccoymerrell committed:
Update 407-writeback-dirty.cc

maccoymerrell commented:
Noticed these test failures. Was unable to use a config file with any new version of develop. ./config.sh would return some sort of error. Default build works fine. In previous version of develop where config.sh didn't error out, it failed to use the prefetcher specified in the config file. However, replacement policy was used.

maccoymerrell committed:
Merge pull request #1 from ChampSim/master

Fixed bug with cache line dirty flag being reset. (#389)

maccoymerrell committed:
Merge pull request #2 from ChampSim/develop

Develop

maccoymerrell committed:
Added Ramulator Support partially

maccoymerrell committed:
Refactored, corrected non-json printouts

maccoymerrell committed:
Added comments on future work

maccoymerrell committed:
Merge branch 'develop' into DRAM_Controller

maccoymerrell committed:
Updated to ramulator2

maccoymerrell committed:
Fixed frequency mismatch. Added comments for future work.

maccoymerrell committed:
Added ramulator config option to .config file

maccoymerrell commented:
To explicitly state, new option is "model" under "DRAM". If "model" is set to "ramulator", then "ramulator_config" should also be defined, which points to a ramulator config file (with extension .yaml).
This implementation is for ramulator2, rather than original ramulator

maccoymerrell commented:
To simplify the process, ChampSim links the dynamic library of ramulator2. This means that ramulator2 builds separately, and system paths will need to point to the library's location when running ChampSim.

maccoymerrell committed:
Added necessary packages for Ramulator2

maccoymerrell commented:
Added necessary ramulator libraries to vcpkg. Any additional ramulator components introduced into champsim (such as a stat tracker) will need them to build properly due to header file conflicts. C++20 will also be required, which currently causes issues with std::string_view objects and fmt. In its current state, it will compile with unaltered ramulator2 without these being necessary as long as no ramulator extensions are created to compile alongside champsim.

maccoymerrell committed:
Added assertions to ensure future ramulator compatability

maccoymerrell committed:
Slight change made to reduce compile warning messages.

maccoymerrell committed:
Force ChampSim frontend, clock ratios

maccoymerrell committed:
Merge branch 'develop' into DRAM_Controller

maccoymerrell commented:
Merged, will likely result in issues with Ramulator-integration. Further testing needs to be done before pull.

maccoymerrell committed:
Fixed merge issues, corrected additional tests for new DRAM timing.

maccoymerrell committed:
Change made to plain printer to avoid complications due to switch to time domain.

maccoymerrell committed:
Updated Ramulator integration to be less convoluted.

maccoymerrell committed:
Added translation status check to matching vpages when finishing translation

This prevents physical address overwrites for packets that are already translated - which can be an issue during PTW walks.

maccoymerrell committed:
Added check for translation status when matching vpages on translation finishes.

This prevents translation finishes from overwriting already-translated packets. This prevents rare behavior where PTW walks' physical addresses are overwritten, resulting in deadlock.

maccoymerrell committed:
Removed commented-out condition on debug statement

maccoymerrell committed:
Removed additions to debug statement

maccoymerrell committed:
Fixed formatting

maccoymerrell committed:
updated to new address type.

maccoymerrell committed:
Create 415-translation-collision.cc

Added test for translation merge behavior, authored by ngober.

maccoymerrell committed:
Merge branch 'develop' into PTW-Deadlock-Bug-Fix

maccoymerrell committed:
Merge branch 'develop' into DRAM_Controller

maccoymerrell committed:
Added ramulator2 as submodule, fixed merge issues from develop.

maccoymerrell committed:
Merge remote-tracking branch 'origin/PTW-Deadlock-Bug-Fix' into DRAM_Controller

maccoymerrell committed:
Merge branch 'ChampSim:master' into DRAM_Controller

maccoymerrell committed:
Merge remote-tracking branch 'upstream/develop' into DRAM_Controller

maccoymerrell committed:
Merge branch 'DRAM_Controller' of https://github.com/maccoymerrell/ChampSimFix into DRAM_Controller

maccoymerrell committed:
Integrated refresh configs into config file.

maccoymerrell committed:
Created a framework for adding plugins to ramulator,  and champsim makefile can now compile ramulator when needed.

maccoymerrell committed:
Merge branch 'develop' into DRAM_Controller

ngober reviewed (changes requested):
I like this patch a lot. My primary concerns have to do with the way that everything gets compiled in. The method you suggest relies pretty heavily on a global make variable and `#ifdef RAMULATOR`. I think we should instead use Make's dependencies and to separate the ramulator patching into a separate file.

ngober committed:
Merge branch 'develop' into DRAM_Controller

maccoymerrell committed:
Merge branch 'develop' into DRAM_Controller

maccoymerrell commented:
I am making some changes to address your requests.

For some of these issues, I think it heavily relies on whether it should be optional to include ramulator (at least its compilation) in Champsim.

The main hurdles I am encountering are:
1. Ramulator has a lot of additional dependencies, including CMake. Is it reasonable to include these in vcpkg by default?
2. Switching between the ramulator and champsim dram controllers. I would like to do this precompile if possible, but I think that will require some larger modifications to the python scripts if we avoid using #ifdefs. Do you think thats the most reasonable way to go? I can split the ramulator/champsim dram controller into two separate files, but this will complicate compilation. Should all this switching be delegated to how the python script makes the build directories?
3. What is the best way to get around using a global make variable? Right now, it conveys to the makefile as well as gcc during compilation that we are using Ramulator. The makefile can be altered somewhat, but I am not sure how to know whether ramulator is being used or not without this. The library is only needed at the last step of compilation, how do I trigger make to see it as a dependency beforehand? How can this be controlled by the config file?

maccoymerrell committed:
Temporary commit. Split dram controller into champsim and ramulator files. Redid makefile to clean up ramulator2 integration. Does not compile as-is

maccoymerrell committed:
Merge branch 'DRAM_Controller' of https://github.com/maccoymerrell/ChampSimFix into DRAM_Controller

maccoymerrell committed:
Merge branch 'develop' into DRAM_Controller

maccoymerrell commented:
I expect these tests to fail, just bringing in all the changes from develop.

maccoymerrell committed:
Fixed 'most' of the issues stemming from merger with develop.

maccoymerrell committed:
Patched tests to work with new microsecond refresh_period conversion on dram controller.

maccoymerrell committed:
Fixed multiconfig. May need more changes for ramulator builds

maccoymerrell commented:
To do:
 - Clean up ramulator integration. This will probably involve either treating the dram controller entirely like a module, which will be a lot of work, or maybe trying to cram everything into a make recipe within the make file so users call make ramulator when building with ramulator instead of using the config file to specify it. The former will probably be necessary for multiconfig support.
 
 - Right now ramulator config is controlled by a separate file. This is probably the easiest way to handle it. Ramulator uses device-matching instead of the more customizable approach of ChampSim, meaning you can't vary # of rows, columns, or general size of each device. Similarly, refresh rates, rows per refresh, and frequency are all not directly configurable. Maybe a warning should print to tell the user that the configs in the champsim_config.json are being ignored?
 
 - Proper stat logging and configuration reporting to the rest of ChampSim. This will involve writing some ramulator modules and other logging inside the ramulator side to get the stats that default ChampSim reports.
 
 - Tests for ramulator connection with ChampSim. This will be limited to ensuring packets can go in and out of ramulator, that it matches the desired configuration, and that timings appear accurate. Testing via the ramulator team should cover any other issues that might arise.

maccoymerrell committed:
Maybe fix MacOS problem?

ngober commented:
Some thoughts about your todos:

 - I don't think there's any problem with a `make ramulator` invocation. My concerns previously were with tying all of the build to `make -lramulator`, which is, you know, not what it's making. In particular, we could then make certain files depend on `ramulator` in the Makefile, and it will build automatically.
 - Ramulator config in a separate file isn't a deal-breaker. It might be interesting if the configuration parsed into a form that was readable by ramulator. Or, when the user specifies `"model": "ramulator"`, we can require `"config": "path/to/config"`.
 - You might ping @GinoAC (or I'll do it) about the broken MacOS test. I think he looked at it recently.

maccoymerrell committed:
Cleaned up portions of the ramulator build process.

Simplified ramulator integration into memory controller.

Added ChampSimStats plugin to ramulator.

Added custom ramulator files to force static library compilation instead of shared.

maccoymerrell commented on config/instantiation_file.py:
```diff
@@ -370,8 +372,11 @@ def get_instantiation_lines(cores, caches, ptws, pmem, vmem, env):
+            _rows_per_refresh=int(pmem['rows_per_refresh']),
             _turn_around_time=int(1000*pmem['turn_around_time']),
             _ulptr=vector_string('&'+v for v in upper_levels[pmem['name']]['upper_channels']),
+            _ramulator_config=', ' + '"' + pmem['ramulator_config'] + '"' if pmem['model'] == 'ramulator' else "",
```
It can be relative, from the perspective of the run directory, which is probably expected behavior, so is fine.

maccoymerrell commented on src/vmem.cc:
```diff
@@ -43,9 +43,14 @@ VirtualMemory::VirtualMemory(champsim::data::bytes page_table_page_size, std::si
   }
+
+  #ifdef RAMULATOR
+  fmt::print("WARNING: Ramulator's physical memory size may be smaller than virtual memory size.\n");
```
No, need to come up with way to determine ramulator dram size from config file. :smile:

maccoymerrell committed:
small changes to how stats are passed to dram controller. Will hopefully allow for ramulator2 stat passing.

maccoymerrell committed:
Stats now passed from Ramulator to Champsim. Congested cycles needs adjusted.

maccoymerrell committed:
Ramulator can now decode packets according to its address mapping on the ChampSim side.

maccoymerrell committed:
Added code to force ChampSimPlugin to be active in Ramulator, regardless of config.
Added code to track Ramulator progress inside ChampSim

maccoymerrell committed:
Readded last of removed stats back into ChampSimRamulator
Corrected bug in DRAM sizing
Channel width, internal prefetch size, now adjust properly with Ramulator Config

maccoymerrell committed:
Added testing for ramulator, updated makefile, fixed integrated controller size bug, added ramulator tests to github workflows.

maccoymerrell committed:
Moved yamls to separate directory to fix workflows, added ramulator to clean (removed build directory)

maccoymerrell committed:
Fixed issue with building stand-alone ramulator test executable

maccoymerrell committed:
Fixed makefile, specifies generic cmake if vcpkg didn't need to install one

maccoymerrell committed:
Fixed config directory typos.

maccoymerrell committed:
Fixed compile-only typo on ramulator config test directory

maccoymerrell committed:
Added coverage exclusions to ramulator2 directory

maccoymerrell commented:
I have added ramulator checks to the github workflows.

This seems like a good checkpoint in terms of progress.
This is how things work as of the moment.

- ChampSim setup is identical up to the moment of make.
- If you plan to use ramulator, make ramulator should be used instead of make, and make ramulator-test to test.
- The config json file specifies a dram model, either "integrated" or "ramulator". It follows that if you try to compile with the ramulator model and don't use make ramulator, it won't work. make ramulator will work with both models.
- Tests were added that give some coverage to ramulator integration itself. Certain tests are excluded when running with the ramulator model (primarily: refresh, scheduling, channel sizes, and addressing [all of these are inside ramulator])

In addition, stats are now integrated between ramulator and champsim, reporting should be as equivalent as possible. Data bus congestion cannot be measured precisely in ramulator without changing the source code, so instead we measure how long packets must wait on average before performing their operation and returning (subtracting time spent opening and closing rows).

As far as features outside of ramulator, there is a small fix to how dram size is reported. It is now reported in bytes available on each channel which _should_ be accurate. This is checked with a new test (700). An additional test was added that checks to make sure packets are returned from the dram controller if a return is requested.

Refreshes are incorporated into the integrated controller as well. There is a test (702 I think) that checks that refreshes are issued at the defined refresh period. Refresh period is defined in ms in the config file and passed to ChampSim in microseconds.

maccoymerrell committed:
Merge branch 'ChampSim:master' into master

maccoymerrell committed:
Made more elegant way of managing multiple Memory Controller/Ramulator Instances
Added makefile recipe ramulator-clean
Added support for additional ramulator modules

maccoymerrell committed:
Merge remote-tracking branch 'upstream/develop' into DRAM_Controller

maccoymerrell committed:
Fixed duplicate declaration

maccoymerrell committed:
Fixed SPP, replaced python ref with python3, removed leftover #ifdef

ngober commented:
Requesting a Gino review on this also, since it's a big new thing. I'm excited to review this. I'll let you know what I think by the end of the week.

GinoAC commented:
Very nice! I should be able to take a look over the weekend and get back to you Monday.

maccoymerrell committed:
Final cleanup before pr review

maccoymerrell committed:
Fixed always-true return on push to dram_queue in ramulator

maccoymerrell committed:
Merge branch 'develop' into DRAM_Controller

maccoymerrell committed:
Merge branch 'develop' into DRAM_Controller

ngober commented on Makefile:
```diff
@@ -13,13 +13,25 @@ override BRANCH_ROOT += $(addsuffix /branch,$(MODULE_ROOT))
 override BTB_ROOT += $(addsuffix /btb,$(MODULE_ROOT))
 override PREFETCH_ROOT += $(addsuffix /prefetcher,$(MODULE_ROOT))
 override REPLACEMENT_ROOT += $(addsuffix /replacement,$(MODULE_ROOT))
+#allows for swappable dram controller (in a future commit, this should probably become a true module)
+override DRAM_CONTROLLER_ROOT += $(addsuffix /$(DRAM_MODEL),$(addsuffix /dram_controller,$(MODULE_ROOT))) 
```
First of all, I'm _very_ dubious that the controller ought to be a true module. The existing modules are all things that the user might write themselves and swap in/out for experiments. This is not that. People have disagreed with me on this point, though, particularly when we were talking about inclusive/exclusive caches. I'm willing to be convinced here, but the mechanisms for modularity are so clunky that I'm hesitant to add more of them.

Also, I don't think I like the design of selectively compiling based on path name here. In my experience, we should compile **all** of the code **every** time. Otherwise, code will languish and break from disuse over time. See the work that I did converting the debug prints from `#ifdef` (selectively compiling source code) to `if constexpr` (compiled, then removed). If we build in selective compilation here, it will lead to maintenance headaches in the future.

maccoymerrell replied:
When I decided to try to integrate Ramulator, I knew in my mind that I shouldn't tie ChampSim's compilation to Ramulator. We take all the baggage of Ramulator's compilation and add it to ChampSim.
Note that the workflows I am running for Ramulator, only a few of the compilers are run. If there isn't a way to separate them, we can't compile with those excluded compiler versions either.

If we are talking in terms of a piece of code we compile in but selectively use, then that sounds a lot like how the module system currently works to me.

GinoAC replied:
I think the main benefit from making the dram controller a module is that we could lay the groundwork for introducing a network on chip interface, but that's probably best left for a different patch and in-depth discussion.

ngober replied:
I agree that there's potential for a future direction, but I think we should keep our changes minimal until we can clarify the requirements for such a change. See my recent thinking [here](https://github.com/ChampSim/ChampSim/issues/424#issuecomment-2353592540).

ngober reviewed (changes requested):
I'm really glad to see this patch. I've hoped to see something like it.

I have some design concerns in the comments that I think need to be addressed. I think the most significant is this: I don't like that there are multiple definitions of `MEMORY_CONTROLLER` in the codebase, one under `dram_controller/integrated` and the other under `dram_controller/ramulator`. You are distinguishing between them in the build script, choosing one directory over another. This will absolutely lead to a problem somewhere along the line when one file gets modified and the other doesn't, or even when a new user looks at them and wonders how this even compiles correctly. I think that needs some discussion.

Thanks for your hard work on this!

maccoymerrell committed:
Update dram_controller/integrated/dram_controller.cc

Co-authored-by: Nathan Gober <ngober@protonmail.com>

maccoymerrell committed:
Update dram_controller/integrated/dram_controller.cc

Co-authored-by: Nathan Gober <ngober@protonmail.com>

maccoymerrell commented on dram_controller/integrated/dram_controller.cc:
```diff
@@ -129,6 +132,51 @@ long DRAM_CHANNEL::finish_dbus_request()
+      b_req.need_refresh = true;
+    }
+    //refresh is being scheduled for this bank
+    if(b_req.need_refresh && !b_req.valid)
```
There are three steps to actually scheduling refreshes.
1. Refresh time needs to have come
2. The bank needs marked for refresh
3. The bank is not busy

This line makes sure the bank is not busy before occupying it with a refresh. We can't refresh a bank that is in use.
Refreshes must compete with requests coming in and out of DRAM. This gives strict priority to refresh, which is correct.

ngober replied:
I see, the answer to my question is yes because banks in use cannot be refreshed.

maccoymerrell commented on test/cpp/src/702-dram-refresh.cc:
```diff
@@ -0,0 +1,101 @@
+        cycles++;
+        //make sure that for every refresh cycle, each bank undergoes refresh at least once
+        std::vector<bool> bank_under_refresh;
+        std::transform(std::begin(uut->channels[0].bank_request),std::end(uut->channels[0].bank_request),std::back_inserter(bank_under_refresh),[](const auto& entry){return(entry.under_refresh);});
```
The problem is two-fold for Ramulator.
First, we can't see inside Ramulator easily. The ChampSimPlugin I wrote gathers as much data as it can.
The second is that Ramulator doesn't actually track the row under refresh. This is done internally in the DRAM chips typically, so the simulation doesn't _need_ to know it.

What we could do it grab the REFRESHES_ISSUED stat from ChampSimPlugin, and at least guarantee that the expected number of refreshes happened within the period. Though this will involve a lot of #ifdef splicing to pull off.

ngober replied:
How does refreshing show externally? Is the bank stalled until the refresh happens, perhaps? Let's write tests to observe things like that.

maccoymerrell commented on test/cpp/src/703-dram-packet-return.cc:
```diff
@@ -0,0 +1,75 @@
+            std::transform(channel_uut.returned.begin(), channel_uut.returned.end(), std::back_inserter(actual_returns), [](champsim::channel::response_type r){return(r.address.to<uint64_t>());});
+            THEN("The packets are returned on the upstream channel")
+            {
+                REQUIRE_THAT(actual_returns, Catch::Matchers::UnorderedEquals(expected_returns));
```
We are not testing the internal scheduling of the memory controller here, so I felt that would be doing what test 701 does already. This ensures that packets are actually being returned when we request, and not if we didn't request a return. This test fills in a gap in our coverage specifically aimed at Ramulator, since we can't run test 701 using it.

maccoymerrell commented on test/cpp/src/750-channel-dimensions.cc:
```diff
@@ -1,3 +1,5 @@
+
+#ifndef RAMULATOR_TEST
```
We could potentially make it work, but the test itself will need to change a lot. Specifically, we will need to instantiate a MEMORY_CONTROLLER instead of individual channels. Ramulator simulates from the memory controller level, so there is one instance of Ramulator and it runs from the memory controller instead of the channels.

maccoymerrell commented on Makefile:
```diff
@@ -196,7 +210,33 @@ ifeq (,$(filter clean configclean pytest maketest, $(MAKECMDGOALS)))
+	@-$(RM) $(executable_name)
+	$(MAKE) ramulator-exec
+
+ramulator-clean:
+	cd ramulator2 && \
+	cd build && \
+	$(MAKE) clean;
```
I wasn't sure whether we should tie the two cleans together. In my mind, I didn't want anything from Ramulator trying to run at all if the user didn't explicitly call for it.

ngober replied:
I think clean is harmless, and I think tying them together gets the behavior the user wants. When they say clean, they mean clean, and we're supposed to know how to do that.

maccoymerrell commented on Makefile:
```diff
@@ -196,7 +210,33 @@ ifeq (,$(filter clean configclean pytest maketest, $(MAKECMDGOALS)))
+	$(CMAKE_EXE) .. && \
+	$(MAKE);
+
+FORCE: ;
```
Probably.

ngober replied:
What I mean is that I've never seen this before. Is it an idiom?

maccoymerrell replied:
It is supposed to force the recipe to run, even if the target already exists. Idk,

maccoymerrell commented on Makefile:
```diff
@@ -274,11 +316,21 @@ $(test_main_name): override CPPFLAGS += -DCHAMPSIM_TEST_BUILD
+
+#For building executable with ramulator
+ramulator-exec: override LDFLAGS += -L$(RAMULATOR_ROOT)/build
+ramulator-exec: override LDLIBS += -lspdlog -Wl,--whole-archive -lramulator -Wl,--no-whole-archive -lyaml-cpp
```
libramulator.a isn't actually added to the objects list in the Makefile, so I needed to link it in somehow.

ngober replied:
Ah, after looking through [the GCC docs](https://gcc.gnu.org/onlinedocs/gcc/Link-Options.html#index-Libraries), it looks like you can either pass the static library in the same way as the object files or with `-l`.

maccoymerrell committed:
Update dram_controller/integrated/dram_controller.cc

Co-authored-by: Nathan Gober <ngober@protonmail.com>

maccoymerrell commented:
@ngober Then I think the thing preventing this pull is the compilation itself. 

I think I could manage to wrap all Ramulator calls as generic functions that don't require ramulator to be present in the system to compile, and if Ramulator is present, somehow override those functions to connect into ramulator? That would allow ChampSim compilation to be possible without Ramulator, and then we could assert Ramulator is present internally when using the ramulator model.
This also solves the problem of selective compilation, since both controllers would be present in the source code.

ngober commented:
Virtual functions are probably fine here. You could also look into PIMPL and `#if __has_include()` as ways to do it. I would really like as much of the simulator to be compiled as possible every time. I believe it's the best long-term strategy.

GinoAC commented:
The weekend did not offer me any time to look at this, I'll be reviewing it throughout the week but I'm pretty excited about what I've seen so far. Thanks for contributing!

maccoymerrell commented:
See pr #504, I have pulled non-ramulator based changes into that pr.

maccoymerrell committed:
Merge branch 'develop' into DRAM_Controller

iamryanhu committed:
Fix TEST_NUM parameter parsing in Makefile

  The TEST_NUM parameter was not working correctly when running individual
  tests due to incorrect pattern matching in the selected_test variable.
  This prevented commands like 'make test TEST_NUM=700' from executing.

  Fixed by correcting the addprefix/filter/addsuffix pattern matching
  logic to properly identify test files based on the provided TEST_NUM.

  Before: make test TEST_NUM=700 would fail to find matching tests
  After: make test TEST_NUM=700 correctly runs test-700-*.cc

  Tested with various TEST_NUM values and TEST_ORDER parameters.

ngober committed:
Merge pull request #651 from iamryanhu/fix/makefile-TESTNUM

Fix TEST_NUM parameter parsing in Makefile

maccoymerrell committed:
Merge branch 'master' of https://github.com/ChampSim/ChampSim

maccoymerrell committed:
Merge branch 'master' of https://github.com/maccoymerrell/ChampSimFix into DRAM_Controller

maccoymerrell committed:
Removed ramulator2 directory

maccoymerrell committed:
Added ramulator2 as submodule

maccoymerrell committed:
Fixed config.json issue

maccoymerrell committed:
Mixed merge issues.

maccoymerrell committed:
Merge branch 'develop' into DRAM_Controller
